File: AgricultureInspection.java
import java.util.ArrayList;
import java.util.List;

// Інтерфейс для дій з інспекціями
interface InspectionRequest {
    void publishRequest();
}

// Абстрактний клас для учасників
abstract class Participant {
    private String name;

    public Participant(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }

    public abstract void displayInfo();
}

// Клас Фермер
class Farmer extends Participant implements InspectionRequest {
    public Farmer(String name) {
        super(name);
    }

    @Override
    public void publishRequest() {
        System.out.println(getName() + ": Публікує запит на інспекцію.");
    }

    @Override
    public void displayInfo() {
        System.out.println("Фермер: " + getName());
    }
}

// Клас Оператор дронів
class Operator extends Participant {
    public Operator(String name) {
        super(name);
    }

    public void receiveRequest() {
        System.out.println(getName() + ": Приймає запит на інспекцію.");
    }

    public void buildRoute(Drone drone) {
        System.out.println(getName() + ": Будує маршрут для дрона.");
        drone.performInspection();
    }

    @Override
    public void displayInfo() {
        System.out.println("Оператор: " + getName());
    }
}

// Клас Дрон
class Drone {
    public void performInspection() {
        System.out.println("Дрон: Виконує інспекцію поля.");
    }

    public void sendData(Operator operator) {
        System.out.println("Дрон: Дані інспекції отримано.");
    }
}

// Клас Агроном
class Agronomist extends Participant {
    public Agronomist(String name) {
        super(name);
    }

    public void analyzeData() {
        System.out.println(getName() + ": Аналізує дані.");
    }

    public void sendAnalysis(Farmer farmer) {
        System.out.println(getName() + ": Відправляє аналітику фермеру.");
    }

    @Override
    public void displayInfo() {
        System.out.println("Агроном: " + getName());
    }
}

// Головний клас для виконання
public class AgricultureInspection {
    public static void main(String[] args) {
        Farmer farmer = new Farmer("Іван");
        Operator operator = new Operator("Олексій");
        Drone drone = new Drone();
        Agronomist agronomist = new Agronomist("Марія");

        // Процес інспекції
        farmer.publishRequest();
        operator.receiveRequest();
        operator.buildRoute(drone);
        drone.sendData(operator);
        agronomist.analyzeData();
        agronomist.sendAnalysis(farmer);
        farmer.displayInfo();
        System.out.println("Інспекція завершена, дані отримано.");
    }
}

File: AgricultureInspection.cs
using System;

// Интерфейс для действий с инспекциями
interface IInspectionRequest
{
    void PublishRequest();
}

// Абстрактный класс для участников
abstract class Participant
{
    public string Name { get; private set; }

    protected Participant(string name)
    {
        Name = name;
    }

    public abstract void DisplayInfo();
}

// Класс Фермер
class Farmer : Participant, IInspectionRequest
{
    public Farmer(string name) : base(name) { }

    public void PublishRequest()
    {
        Console.WriteLine($"{Name}: Публикует запрос на инспекцию.");
    }

    public override void DisplayInfo()
    {
        Console.WriteLine($"Фермер: {Name}");
    }
}

// Класс Оператор дронов
class Operator : Participant
{
    public Operator(string name) : base(name) { }

    public void ReceiveRequest()
    {
        Console.WriteLine($"{Name}: Принимает запрос на инспекцию.");
    }

    public void BuildRoute(Drone drone)
    {
        Console.WriteLine($"{Name}: Строит маршрут для дрона.");
        drone.PerformInspection();
    }

    public override void DisplayInfo()
    {
        Console.WriteLine($"Оператор: {Name}");
    }
}

// Класс Дрон
class Drone
{
    public void PerformInspection()
    {
        Console.WriteLine("Дрон: Выполняет инспекцию поля.");
    }

    public void SendData(Operator operatorInstance)
    {
        Console.WriteLine("Дрон: Данные инспекции получены.");
    }
}

// Класс Агроном
class Agronomist : Participant
{
    public Agronomist(string name) : base(name) { }

    public void AnalyzeData()
    {
        Console.WriteLine($"{Name}: Анализирует данные.");
    }

    public void SendAnalysis(Farmer farmer)
    {
        Console.WriteLine($"{Name}: Отправляет аналитику фермеру.");
    }

    public override void DisplayInfo()
    {
        Console.WriteLine($"Агроном: {Name}");
    }
}

// Главный класс для выполнения
class AgricultureInspection
{
    static void Main(string[] args)
    {
        Farmer farmer = new Farmer("Иван");
        Operator operatorInstance = new Operator("Алексей");
        Drone drone = new Drone();
        Agronomist agronomist = new Agronomist("Мария");

        // Процесс инспекции
        farmer.PublishRequest();
        operatorInstance.ReceiveRequest();
        operatorInstance.BuildRoute(drone);
        drone.SendData(operatorInstance);
        agronomist.AnalyzeData();
        agronomist.SendAnalysis(farmer);
        farmer.DisplayInfo();
        Console.WriteLine("Инспекция завершена, данные получены.");
    }
}
